### ISSUE TITLE
Extremely slow toy fits with zfit compared with ROOFIT

### ISSUE THREAD

**Issue by acampove** [zfit/zfit#664]
## Current Behaviour

I am running a simple Gaussian plus exponential toy test and I see that running over 1000 toys takes around 3 and a half minutes with zfit. With ROOFIT, with the same settings, I can get this done in six seconds. I seem to see a difference of two orders of magnitude.

The reproducers in C++ (ROOFIT) and python are:

```python
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

import tqdm
import zfit

obs     = zfit.Space('x', limits=(0, 10))
mu      = zfit.Parameter('mu', 5.0, 0, 10)
sg      = zfit.Parameter('sg', 1.0, 0, 10)
gaus    = zfit.pdf.Gauss(obs=obs, mu=mu, sigma=sg)
nsig    = zfit.Parameter('nsig', 1000, 0, 10_000)
gaus    = gaus.create_extended(nsig)

lam     = zfit.Parameter('lb', -0.01, -1, 0)
exp     = zfit.pdf.Exponential(lam=lam, obs=obs)
nbkg    = zfit.Parameter('nbkg', 1000, 0, 10_000)
exp     = exp.create_extended(nbkg)

model   = zfit.pdf.SumPDF([gaus, exp]) 
sampler = model.create_sampler(n=2000)

nll     = zfit.loss.ExtendedUnbinnedNLL(model=model, data=sampler)
minimizer = zfit.minimize.Minuit()

nruns = 1000
for run_number in tqdm.tqdm(range(nruns), ascii=' -'):
    mu.set_value(5.0)
    sg.set_value(1.0)

    sampler.resample()
    res = minimizer.minimize(nll)
```

```c++
#include "RooRealVar.h"
#include "RooDataSet.h"
#include "RooGaussian.h"
#include "RooChebychev.h"
#include "RooAddPdf.h"
#include "RooMCStudy.h"
#include "RooPlot.h"
#include "TCanvas.h"
#include "TAxis.h"
#include "TH2.h"
#include "RooFitResult.h"
#include "TStyle.h"
#include "TDirectory.h"
 
using namespace RooFit;
 
void rf801_mcstudy()
{
   RooRealVar x("x", "x", 0, 10);
   x.setBins(40);
 
   // Create two Gaussian PDFs g1(x,mean1,sigma) anf g2(x,mean2,sigma) and their parameters
   RooRealVar mean("mean", "mean of gaussians", 5, 0, 10);
   RooRealVar sigma("sigma", "width of gaussians", 1.0, 0.0, 10.0);
   RooGaussian sig("sig", "Signal component 1", x, mean, sigma);
 
   RooRealVar lm("lm", "lm", -0.01, -1, 0.);
   RooExponential bkg("bkg", "Background", x, lm);
 
 
   RooRealVar nbkg("nbkg", "number of background events,", 1000, 0, 10000);
   RooRealVar nsig("nsig", "number of signal events", 1000, 0, 10000);
   RooAddPdf model("model", "model", RooArgList(bkg, sig), RooArgList(nbkg, nsig));
 
   RooMCStudy *mcstudy =
      new RooMCStudy(
              model, 
              x, 
              Binned(false), 
              Silence(), 
              Extended(), 
              FitOptions(Save(true), PrintEvalErrors(0)));
 
   mcstudy->generateAndFit(1000);

   RooPlot *frame = mcstudy->plotPull(mean, Bins(40), FitGauss(true));
   gStyle->SetOptStat(0);
   TCanvas *c = new TCanvas("rf801_mcstudy", "rf801_mcstudy", 900, 900);
   frame->Draw();
}
```

These should be doing mostly the same, with just the underlying fitting code been the difference. Am I doing something that is terribly inefficient in the zfit example?

## Expected Behaviour

The zfit version should be within a factor of 2.

## Context (Environment)

- zfit version: 0.27.1
- Python version: 3.12.11
- Are you using conda, pipenv, etc? : micromamba
- Operating System: AlmaLinux9
- Tensorflow version: 2.19.0

**Comment by acampove** [comment]
In order to run the roofit version do:

```bash
time root -l -q rf801_mcstudy.cxx
```

in the terminal.

**Comment by acampove** [comment]
After some experimentation it seems that:

```
minimizer = zfit.minimize.Minuit(
    gradient='zfit',
    mode=0)
```

make it faster. `Mode = 0` seems to skip calculating the Hessian. However I am not sure why we would need the Hessian while the minimization happens. That should only be needed for the error calculation.

I also see that I need to turn off the GPU for the fit to run faster. Which I do not undertstand, I would naively think the GPU would speed this up.

I seem to go from 2.9 fits per second to 9 fits per second. Which is not 2 orders of magnitude so there must be something else happening.

For instance, I see that the resampling step alone (no minimization) runs at 33 iterations per second. That step alone would make this 4 times slower than ROOFIT.

**Comment by jonas-eschle** [comment]
Hi, thanks a lot for bringing this up! The result you get in the end is actually somewhat expected, but let me disect things:

> These should be doing mostly the same, with just the underlying fitting code been the difference.

Not quite, this is probably a false premise. zfit and RooFit, while similar on the top layer, differ significantly under the hood. 
zfit offers two "advantages" over RooFit: it's pure Python, which makes a lot of steps easier (i.e. integration with other libraries, implementing something hacky. It also is scalable, and this is nowadays an "advantage" in quotation marks, because RooFit did a lot of improvements with GPU support etc. But the main point is, that zfit, due to it's hackability and performance for larger fits, including the fact that an external computational backend is used, was never meant to be fast for _small fits_. Because they often don't matter. Sure, it's nice if toy fit's take 10 secs instead of a few mins, no doubt, but it will never be a blocker. If a fit takes 10 mins (a single one) or 10 hours, that matters. zfit is clearly aiming at the latter case in terms of speed.

These are advantages and disadvantages, both expected. The goal in terms of speed for zfit is to be within an order of magnitude of RooFit, for really small fits, that can be more, for larger fits, it should be within a factor of 2-3 rather.

If you take the example and instead of 2K use 200K, they are basically at the same speed (with your improvements). Not saying they're exactly the same, but basically.

> Mode = 0 seems to skip calculating the Hessian. However I am not sure why we would need the Hessian while the minimization happens. That should only be needed for the error calculation.

That's a minuit thing. If you want to check the converge, you need a good criterion. minuit uses the EDM, which uses the hessian (and zfit, in general, uses it also for other minimizers). Using mode=0, the internal hessian is not explicitly calculated (but that's all a minuit feature, feel free to read the manual or see on iminuit, there is quite some information around it (and play around)).

> I also see that I need to turn off the GPU for the fit to run faster. Which I do not undertstand, I would naively think the GPU would speed this up.

That's very well expected, you can also see the (very old, very specific) benchmark in the zfit paper: 

<img width="1500" height="1094" alt="Image" src="https://github.com/user-attachments/assets/63a5f84c-15ee-4dc8-b348-30f9a4c644ad" />

moving data to the GPU is a bottleneck, so the NLL is evaluated on the GPU but the number needs to go back to the CPU for the minimizer. If the evaluation of the likelihood is really simple (i.e. only a few thousand events), moving this to the GPU, evaluating it is simply not faster than just calculating it on the CPU. But this is a general thing, you can also read up on GPU vs CPU calculations, as an example, see also cupy tutorials or similar. Often, for only a few K events, using the CPU is a lot faster.
So no, GPU is not for free and only worth if you have ~100K events in my experience (or maybe 20K+)

But I am also benchmarking things, always good to see if there is room for improvements

**Comment by acampove** [comment]
Hi @jonas-eschle 

In most cases, our datasets are a few thousand events. Because we look for rare processes, that in many cases haven't been observed. So I would not think of the low statistics case as a corner case, it's actually the opposite.

I see you refer to [this](https://arxiv.org/pdf/1910.13429) paper. Another plot I would have made would have been one comparing timing vs the number of parameters. The most common use-case is a dataset with maybe 1000-10000 events and a model with 20-50 parameters, where we only have 3-4 unconstrained. That is likely the most common case in real analyses.

Regarding the usage of the Hessian for calculating the EDM, I wonder how needed this is. I see that only `mode = 2` does a proper calculation `mode=1` (the default) is just a quick update based on some BFGS, see [this](https://indico.cern.ch/event/1279438/contributions/5383385/attachments/2640182/4568768/Minuit2_CHEP2023.pdf) (unfortunately a proper publication on MIGRAD does not exist). But from what I understand, the `pseudo-Hessian` is not used for the EDM only, but also to find out the size of the step needed.

I see three scenarios:

- **Large statistics sample**:  Calculating the Hessian might make sense in order to reduce the minimization steps.
- **Small statistics sample**:  I am running the fit with a low statistics (1000 samples) dataset and the `pseudo Hessian` approach actually makes this slower
- **Large number of parameters**: Haven't tried this yet, but I believe this would worsen the performance, given that the time spent calculating the hessian grows quadratically with the number of parameters. And [this](https://scikit-hep.org/iminuit/faq.html#what-happens-when-i-change-the-strategy) recommends using `mode=0` for models with more than 10 parameters, i.e. a plain gradient descent.

Which leaves us with only the `large statistics + few parameters` case where this MIGRAD approach might outperform the simpler gradient descent approach. However I haven't tested that myself.

**Comment by acampove** [comment]
The place where this does matter is when we are trying to validate our toys. There are tools that we are building for which we need tests. Those tests:

- Set a random seed for reproducibility
- Build toy models
- Create toy data from those models and fit them

This enables picking up the correlations from a real fit. Also, we create small models to speed up the tests. Tests should not take more than a few seconds in order not to slow down the development. And by making large statistics tests (e.g. 10K entries) we can look at the results and compare them with our expectations beyond the low statistics noise. Thus we can spot more easily and early any issue with our fitting framework. 

If a quick small fit is not possible, then we might have to:

1. Use fake fitted errors and values in the form of `numpy.random.normal` arrays. Not ideal, given that the correlations are lost.
2. Use ROOFit just in the creation of inputs for tests. Not ideal either, I would rather use the same framework all over.
3. Create a model and fit toys only once and save the parameters as part of our package and re-use those toys for debugging purposes. Not ideal, because this will make our package larger in the long term and we would only be testing what happens after the fits, not the full chain with the fits. e.g. if we break the code, that changes the fits, but we are not re-making the fitting parameter list, we are just reusing the old list from fits done with the old code.

Currently I am taking approach No 1. No 3 might be a good idea in the long term. So yeah, this is a case where having fast small fits for testing the framework would help, despite the real fit will be large and time consuming.

**Comment by jonas-eschle** [comment]
> Tests should not take more than a few seconds in order not to slow down the development.

Sorry, the development of what exactly? The library or for the end user? I may missed this point.

Just to note, you could also parallelize in general on a cluster or similar, this also helps

> Currently I am taking approach No 1. No 3 might be a good idea in the long term. 

okay, I see. That's something hepstats started to build in, namely to store toys. If in your case they take seconds, or minutes, that's still very comfortable, but toys that take days should definitely not be lost. So maybe that's anyway a good strategy?

**Comment by acampove** [comment]
Hi @jonas-eschle 

I am back to this, I had been doing other stuff.

> > Tests should not take more than a few seconds in order not to slow down the development.
> 
> Sorry, the development of what exactly? The library or for the end user? I may missed this point.

The end user, when we develop code, we need something fast, if running over 1000 toys for a trivial model (exp + gaus) takes 5 minutes that's going to make debugging anything a nightmare.
> 
> Just to note, you could also parallelize in general on a cluster or similar, this also helps

Right, but then we are going to run what would be local tests in a cluster? I mean the pytest type of tests. I think this might also be a bottleneck for zfit. Because you have tests, that you run with e.g. pytest. Many of those tests do not need to test complex models, a gaussian plus exponential are enough. But to check the health of zfit you still need to get pulls with lets say 1000 toys. So even zfit development is going to be a pain like this.
> 
> > Currently I am taking approach No 1. No 3 might be a good idea in the long term.
> 
> okay, I see. That's something hepstats started to build in, namely to store toys. If in your case they take seconds, or minutes, that's still very comfortable, but toys that take days should definitely not be lost. So maybe that's anyway a good strategy?

The strategy is to profile zfit and find out where the bottleneck is and deal with it. ROOFIT can do these fits about 100 times faster so for sure there must be a big bottleneck somewhere.
